File: staks_and_queues/stack.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Stack:
    def __init__(self):
        self.first = None
        self.last = None
        self.length = 0

    def add(self, value):
        node = Node(value)

        if self.length == 0:
            self.first = node
            self.last = node
        else:
            node.next = self.first
            self.first = node
        self.length += 1

    def pop(self):
        if self.length == 0:
            return None
        pop_node = self.first

        if self.first == self.last:
            self.first = None
            self.last = None
        else:
            self.first = self.first.next

        self.length -= 1

        return pop_node

File: staks_and_queues/stack.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class Stack {
  constructor(){
    this.first = null;
    this.last = null;
    this.size = 0;
  }

  push(value) {
    let node = new Node(value);
    if (this.size === 0) {
      this.first = node;
      this.last = node;
    } else {
      let previousFirst = this.first;
      this.first = node;
      this.first.next = previousFirst;
    }
    this.size++;
    return this.size;
  }

  pop(){
    if(this.size === 0) return null;
    if (this.size === 1) {
      this.first = null;
      this.last = null;
    } else {
      const removedNode = this.first;
      this.first = this.first.next;
      removedNode.next = null;      
    }
    this.size--;
    return removedNode;
  }

}